quickstart/src/main/java/com/microsoft/azure/kusto/quickstart/KustoSampleApp.java

```diff
@@ -0,0 +1,399 @@
+import java.net.URISyntaxException;
+import java.security.GeneralSecurityException;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.*;
+import java.util.concurrent.TimeUnit;
+
+import static com.microsoft.azure.kusto.data.ClientRequestProperties.OPTION_SERVER_TIMEOUT;
+
+public class KustoSampleApp {
+    // TODO - Config:
+    //  If this quickstart app was downloaded from OneClick, kusto_sample_config.json should be pre-populated with your cluster's details
+    //  If this quickstart app was downloaded from GitHub, edit kusto_sample_config.json and modify the cluster URL and database fields appropriately
+    private static final String CONFIG_FILE_NAME = "kusto-sample-config.json";
+
+    // TODO - Config (Optional): Change the authentication method from User Prompt to any of the other options
+    //  Some of the auth modes require additional environment variables to be set in order to work (see usage in generateConnectionString below)
+    //  Managed Identity Authentication only works when running as an Azure service (webapp, function, etc.)
+    private static final String AUTHENTICATION_MODE = "UserPrompt"; // Options: (UserPrompt|ManagedIdentity|AppKey|AppCertificate)
+
+    // TODO - Config (Optional): Toggle to false to execute this script "unattended"
+    private static final boolean WAIT_FOR_USER = true;
+
+    private static String databaseName;
+    private static String tableName;
+    private static String kustoUrl;
+    private static String ingestUrl;
+    private static boolean useExistingTable;
+    private static List<Map<String, String>> dataToIngest;
+    private static String tableSchema;
+    private static int step = 1;
+    private static final String BATCHING_POLICY = "{ \"MaximumBatchingTimeSpan\": \"00:00:10\", \"MaximumNumberOfItems\": 500, \"MaximumRawDataSizeMB\": 1024 }";
+    private static final int WAIT_FOR_INGEST_SECONDS = 20;
+
+    public enum ExecutionType {
+        CONTROL_COMMAND("Control Command"),
+        QUERY("Query");
+
+        private final String name;
+
+        ExecutionType(String name) {
+            this.name = name;
+        }
+
+        public String getName() {
+            return name;
+        }
+    }
+
+    public static void main(String[] args) throws InterruptedException {
+        System.out.println("Kusto sample app is starting...");
+
+        loadConfigs();
+
+        if (AUTHENTICATION_MODE.equals("UserPrompt")) {
+            waitForUserToProceed("You will be prompted *twice* for credentials during this script. Please return to the console after authenticating.");
+        }
+
+        ConnectionStringBuilder kustoCsb = generateConnectionString(kustoUrl, AUTHENTICATION_MODE);
+        ConnectionStringBuilder ingestCsb = generateConnectionString(ingestUrl, AUTHENTICATION_MODE);
+        Client kustoClient = null;
+        try (IngestClient ingestClient = IngestClientFactory.createClient(ingestCsb)) {

```

Comment:
I can also move the ingest client to right before the while loop for ingesting.

The benefit to moving it is to make it clear where it's needed/used.
The benefit to keeping it here is to keep it simple, creating both clients in one place.

I prefer the former, but suspect Yochai prefers the latter.

Reply:
I don't mind actually, though shorten the exception catch to 
```
try{ create client}
catch{exit} 
everything else
```

Reply:
The problem with shortening it is that the **use** of the ingest client needs to be inside the `try`, which is why the catch is so far down.

Reply:
IngestClient = null
Try { 
ingestClient = CreateClient())
}
Catch 
{
error 
}

Do something with the client 

Or do I have java figured out backwards?

Reply:
That's how I did it originally, but the problem with that is it doesn't take advantage of Java's "try-with-resources". It's similar to a `finally`, which ensures that `Closable` resources are properly closed no matter what, and is therefore generally best practices.